src/osd/OSD.cc

```diff
@@ -1479,6 +1479,45 @@ class PGOpQueueable : public OpQueueItem::OpQueueable {
   }
 };
 
+class PGOpQueueableNoLock : public OpQueueItem::OpQueueable {
+  PGRef pg;
+protected:
+  spg_t get_pgid() const {
+    return pg->get_pgid();
+  }
+public:
+  PGOpQueueableNoLock(PGRef pg) : pg(pg) {}
+  virtual uint32_t get_queue_token() const override final {
+    return pg->get_pgid().ps();
+  }
+
+  virtual void *get_ordering_token() const override final {
+    return (pg.get()+1);
+  }
+
+  virtual OpQueueItem::OrderLocker::Ref get_order_locker() override final {
+    class Locker : public OpQueueItem::OrderLocker {
+      PGRef pg;
+    public:
+      Locker(PGRef pg) : pg(pg) {}
+      virtual void lock_suspend_timeout(
+	ThreadPool::TPHandle &handle) override final {
```

Comment:
You need some kind of lock or items queued on the same ordering token can be processed out of order.